### PR TITLE
Model-based tests for ICS02

### PR DESCRIPTION
**Pull request by vitorenesduarte**
Towards: #546

## Description

This PR is the first step towards #601. For ease of reviewing, we plan to introduce one ICS at a time, starting now with ICS02.

For some details on MBT and current limitations, please check [modules/tests/README.md](https://github.com/informalsystems/ibc-rs/blob/vitor/ics02_mbt/modules/tests/README.md) (which steals ideas from the [light-client one](https://github.com/informalsystems/tendermint-rs/blob/master/light-client/tests/support/model_based/README.md)).

______

For contributor use:

- [ ] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/Cargo.toml
Comment (romac) [inline modules/Cargo.toml:40]: This should be moved the `dev-dependencies` section as it is only used in tests and we want to keep the `modules` crate free of error handling libraries.

Comment (vitorenesduarte) [reply modules/Cargo.toml]: Done in d57cf4119a9dbfe5260ef7c1afc212bc5c305458. Once `modelator.rs` lives in its own crate, we will depend only on that, so that dependency can go away.

File: modules/tests/model_based.rs
Comment (adizere) [inline modules/tests/model_based.rs:95]: Chain "epoch" was redefined to "revision" I think, we just didn't get a chance to make it consistent thoughout the codebase. The same goes for the epoch in heights. Ref: https://github.com/informalsystems/ibc-rs/issues/405

True this has nothing to do with `Version` type for channels or connections.

File: modules/tests/model_based.rs
Comment (adizere) [inline modules/tests/model_based.rs:29]: I was wondering if it would have any positive effect if we used the domain type [`ChainId`](https://github.com/informalsystems/ibc-rs/blob/9ee48bd2a7c337277de3889df06d963b0fd005bf/modules/src/ics24_host/identifier.rs#L18) instead of String here. This would make the MBT suite more tedious here because of additional parsing/error catching (though serde could help there), but maybe it would align it better with the rest of the codebase.

File: modules/tests/modelator.rs
Comment (adizere) [inline modules/tests/modelator.rs:1]: Could we give a bit more structure to this directory, e.g., put modelator.rs and step.rs inside a module of their own (which we can call `model_based`).

File: modules/tests/support/model_based/IBC.cfg
Comment (adizere) [inline modules/tests/support/model_based/IBC.cfg:2]: Would be a good idea to give chain identifiers the same format as in production, e.g., "chainA-0" and "chainB-0".

File: modules/tests/support/model_based/IBCTests.cfg
Comment (adizere) [inline modules/tests/support/model_based/IBCTests.cfg:1]: Not clear what is the difference b/t this file and `modules/tests/support/model_based/IBC.cfg`.

File: modules/tests/support/model_based/README.md
Comment (adizere) [inline modules/tests/support/model_based/README.md:1]: I would move this readme to a higher-level, inside `modules/tests/`. Here it's a bit buried and it's easy to overlook it.

**Review by adizere — CHANGES_REQUESTED** [review]
Impressive work!

My main worry is that we're introducing new TLA+ specs here, instead of leveraging (part of) what we already had. I'm curious what @ancazamfir, @andrey-kuprianov, and @istoilkovska think about this: would it be possible, at least eventually, to reuse Ilina's TLA+ specs for MBT?

File: modules/tests/modelator.rs
Comment (romac) [reply modules/tests/modelator.rs:38]: Agreed, it would be nice if we could get away with only `thiserror`.

File: modules/src/ics02_client/msgs/create_client.rs
Comment (romac) [reply modules/src/ics02_client/msgs/create_client.rs:25]: We are quite inconsistent with this all over the codebase, and it's something we should uniformly sort out at some point.

Comment (vitorenesduarte) [reply modules/src/ics02_client/msgs/create_client.rs:25]: The convention seems to be to use `pub`, which explains the change here. I'll do a follow up PR where we remove all the getters from these structs.

File: modules/tests/support/model_based/IBC.cfg
Comment (vitorenesduarte) [reply modules/tests/support/model_based/IBC.cfg:2]: Could you point me to place where this format is defined? Can there be a `"chainA-0"` and a `"chainA-1"`?

File: modules/tests/support/model_based/README.md
Comment (vitorenesduarte) [reply modules/tests/support/model_based/README.md]: Done in 5e9dd634a3bc04c0f193ab3872b83dc00505d1d4.

File: modules/tests/model_based.rs
Comment (vitorenesduarte) [reply modules/tests/model_based.rs]: Done in fb20ce0161ff8ad7009b68af9ff8978e508f4ee9.

File: modules/tests/model_based.rs
Comment (vitorenesduarte) [reply modules/tests/model_based.rs]: Thanks. Renamed `epoch` to `revision` in 75db6bb9621e3a6db269100332e137d3a526a822.

File: modules/tests/modelator.rs
Comment (vitorenesduarte) [reply modules/tests/modelator.rs]: Tried something in a10286398eee2ff375e727720b716d1293bcf12c. Please take a look @adizere @romac. I should also say that the plan is to eventually (hopefully soon) have the functionality in `moderator.rs` moved to its own crate.

File: modules/tests/modelator.rs
Comment (vitorenesduarte) [reply modules/tests/modelator.rs]: Once `modelator.rs` goes way, this directory would only contain `step.rs`. Do you mind if we delay this and try to improve structure in future PRs?

File: modules/tests/modelator.rs
Comment (romac) [reply modules/tests/modelator.rs]: Looks good

File: modules/tests/support/model_based/IBCTests.cfg
Comment (vitorenesduarte) [reply modules/tests/support/model_based/IBCTests.cfg:1]: Tried to improve the README explaining this in 2e5d6de82f69a14046b8555f95c706701b140415.

**Comment by vitorenesduarte** [comment]
Thanks @adizere and @romac for the review! I just finished addressing all your comments.

File: modules/tests/model_based.rs
Comment (ancazamfir) [inline modules/tests/model_based.rs:190]: should we also check that the new client is stored in the `ctx`, i.e. it can be retrieved?

File: modules/tests/model_based.rs
Comment (ancazamfir) [inline modules/tests/model_based.rs:225]: should we also check that the client in `ctx` has the updated consensus state?

File: modules/tests/model_based.rs
Comment (ancazamfir) [inline modules/tests/model_based.rs:243]: should we also check that the client's consensus state has not been updated?

File: modules/tests/step.rs
Comment (ancazamfir) [inline modules/tests/step.rs:29]: For create client we need to have a `client_state` and for update client we need a `header`. It is true that for the mock client these boil down to a `client_height` but if we want to extend these tests for other client types or make changes to the mock client (e.g. add a `chain-id` or client specific parameters to the mock state, or add a timestamp to the header/ consensus state)

**Review by ancazamfir — COMMENTED** [review]
Looks good after a fist pass. Entered a few comments, will take another look.

File: modules/tests/modelator.rs
Comment (andrey-kuprianov) [reply modules/tests/modelator.rs]: yes, minimising the dependencies is a good idea

File: modules/tests/model_based.rs
Comment (vitorenesduarte) [reply modules/tests/model_based.rs:190]: This is a good idea! In the same way we check chain heights after each model action, we can extend it to also check the state of clients. I've opened https://github.com/informalsystems/ibc-rs/issues/629 to track this idea (which is not so easy to implement right now).

File: modules/tests/model_based.rs
Comment (vitorenesduarte) [reply modules/tests/model_based.rs:225]: Same.

File: modules/tests/model_based.rs
Comment (vitorenesduarte) [reply modules/tests/model_based.rs:243]: Same.

File: modules/tests/step.rs
Comment (vitorenesduarte) [reply modules/tests/step.rs:29]: Currently the model does not distinguish between client state or consensus state. For each client, only the current height is kept. For this reason, there's no `"ICS02ConsensusStateNotFound"` error in the model, only `"ICS02ClientNotFound"` (as they cannot be distinguished).

Are you suggesting that the model should be extended to make the distinction? In general, we cannot hope to model everything, but maybe this is one of the situations where we should. At this point it is not obvious to me why the current abstraction is insufficient. It might become clear once there are other client types to be tested.

File: modules/src/ics02_client/msgs/create_client.rs
Comment (vitorenesduarte) [reply modules/src/ics02_client/msgs/create_client.rs:25]: Opened https://github.com/informalsystems/ibc-rs/issues/631.

Comment (adizere) [reply modules/src/ics02_client/msgs/create_client.rs:25]: thank you!

File: modules/tests/support/model_based/IBC.cfg
Comment (adizere) [reply modules/tests/support/model_based/IBC.cfg:2]: This format is just a convention at the moment, not sure it's formally defined anywhere.

> Can there be a "chainA-0" and a "chainA-1"?

This would be the same chain, in two subsequent epochs or revisions: revision `0` (due to suffix `-0`), and revision `1`. This also explains why a ibc::Height has two fields: a revision and a number.

For "chainA-0", the revision would be "0" at all height numbers of that chain.
For "chainA-1", the revision would be "1" at all height numbers of that chain...

File: modules/tests/step.rs
Comment (ancazamfir) [reply modules/tests/step.rs:29]: > Are you suggesting that the model should be extended to make the distinction? 

Yes

> In general, we cannot hope to model everything, but maybe this is one of the situations where we should. At this point it is not obvious to me why the current abstraction is insufficient. It might become clear once there are other client types to be tested.

Understood we cannot model everything. But the create_client message includes a client state (`AnyClientState`) and a consensus state (`AnyConsensusState`) and I think we should pass those in the `action`.  update_client should take `AnyHeader`. 
It's not only for the different client types and MBT testing for clients. When we do MBT for connection we will need to pass `AnyConsensusState` (and not a client height) in some of the messages. And this would be something that we should have previously created a client with (or not, depending on the test).

Comment (vitorenesduarte) [reply modules/tests/step.rs:29]: > Understood we cannot model everything. But the create_client message includes a client state (`AnyClientState`) and a consensus state (`AnyConsensusState`) and I think we should pass those in the `action`. update_client should take `AnyHeader`.

I can make this distinction in the model, but it's not clear to me what it brings to ICS02. For example, what should be a client state and a consensus state in the model? I can model both as a height, but it feels like this is the same as having a single height (which is what's modeled now).

> It's not only for the different client types and MBT testing for clients. When we do MBT for connection we will need to pass `AnyConsensusState` (and not a client height) in some of the messages. And this would be something that we should have previously created a client with (or not, depending on the test).

If the distinction will be needed for ICS03, can we address this concern in the PR with MBT for ICS03?

Comment (ancazamfir) [reply modules/tests/step.rs:29]: We don't have to hold this PR necessarily.  I will leave it up to you. But I think we need to make the change so that we can easily extend the tests. And just to clarify this is not only needed for ICS03.

File: modules/tests/support/model_based/IBC.cfg
Comment (vitorenesduarte) [reply modules/tests/support/model_based/IBC.cfg]: That makes sense, thanks! Done in 5d6bf08c7fcfe4d9040c661f0f044ebf641ea31f.

File: modules/tests/step.rs
Comment (vitorenesduarte) [reply modules/tests/step.rs:29]: After discussing with @ancazamfir, we have reached the conclusion that the model should keep, per client, the latest `N` heights, not just the latest one. The current model is a special case where `N = 1`. I've opened an issue to track this: https://github.com/informalsystems/ibc-rs/issues/633. Thanks @ancazamfir!

**Review by ancazamfir — APPROVED** [review]

**Comment by ancazamfir** [comment]
Let's merge this one and address any pending comments in #633.

**Review by adizere — APPROVED** [review]
Good work Vitor!

**Pull request merged**
